README.md
-   [Introduction](#introduction)
-   [The color scales](#the-color-scales)
-   [Usage](#usage)

<!-- badges: start -->
[![Build
Status](https://api.travis-ci.org/nschiett/fishualize.png?branch=master)](https://travis-ci.org/nschiett/fishualize)
[![CRAN
status](https://www.r-pkg.org/badges/version/fishualize)](https://CRAN.R-project.org/package=fishualize)
[![Downloads](http://cranlogs.r-pkg.org/badges/grand-total/fishualize?color=brightgreen)](https://cran.rstudio.com/package=fishualize)
<!-- badges: end -->

<img src="man/figures/fishualize_logo.png" width = 120 alt="fishflux logo"/>

Introduction
============

The [**fishualize**](http:://github.com/nschiett/fishualize) package
provides color scales for plotting in R based on nature’s most stunning
and colorful organisms: teleost fishes (with a few chondrichthyan
cameos). \#TeamFish in its colorful glory.

Installation
------------

``` r
install.packages("fishualize")
library(fishualize)
```

or for the latest development version:

``` r
library(devtools)
devtools::install_github("nschiett/fishualize", force = TRUE)
library(fishualize)
```

The color scales
================

The package contains one scale per species, defined by five dominant
colors. The number of fishualized species will expand over time. For a
list of fish species that are currently available, run
`fish_palettes()`.  
A visual overview of the color scales can be found
[here](https://nschiett.github.io/fishualize/articles/overview_colors.html).

To visualize a fish color palette, you can run `fishualize()` and
specify your choice.

``` r
fishualize()
```

<img src="README_files/figure-markdown_github/tldr_vis-1.png" width="672" />

``` r
fishualize(n = 8, option = "Hypsypops_rubicundus", end = 0.9)
```

<img src="README_files/figure-markdown_github/tldr_vis-2.png" width="672" />

Usage
=====

The `fish()` function produces the fish color scale based on your
favorite species, which can be specified using ‘option’ =
“Your\_favorite” or fish\_palettes()\[\] with the number of your species
specified.

For base R plots, use the `fish()` function to generate a palette:

``` r
pal <- fish(256, option = "Trimma_lantana")
image(volcano, col = pal)
```

<img src="README_files/figure-markdown_github/tldr_base-1.png" width="672" />

``` r
pal <- fish(10, option = "Ostracion_cubicus")
image(volcano, col = pal)
```

<img src="README_files/figure-markdown_github/tldr_base-2.png" width="672" />

ggplot2
-------

The package also contains color scale functions for **ggplot2** plots:
`scale_color_fish()` and `scale_fill_fish()`.

``` r
library(ggplot2)
library(rfishbase)

# load data for plotting 
# 1. Create list of species names currently featured in fishualize
spp <- fishualize::fish_palettes()
# 2. Get data on the included species from FishBase using the rfishbase package
dt <- rfishbase::species(gsub("_"," ", spp))
```

    ## Warning: `data_frame()` is deprecated as of tibble 1.1.0.
    ## Please use `tibble()` instead.
    ## This warning is displayed once every 8 hours.
    ## Call `lifecycle::last_warnings()` to see where this warning was generated.

``` r
dt$Importance = factor(dt$Importance, levels = c("highly commercial", "commercial", "minor commercial", "subsistence fisheries", "of no interest"))

# plot bars with discrete colors using color scheme provided by Scarus quoyi
ggplot(dt[!is.na(dt$Importance),]) +
  geom_bar(aes(x = Importance, fill = Importance)) +
  scale_fill_fish_d(option = "Scarus_quoyi") +
  theme_bw() +
  theme(axis.text.x= element_blank() )
```

<img src="README_files/figure-markdown_github/tldr_ggplot-1.png" width="672" />

``` r
# plot points with continuous colors provided by Hypsypops rubicundus
ggplot(dt) +
  geom_point(aes(x = Length, y = Vulnerability, color = Vulnerability), size = 3) +
  scale_color_fish(option = "Hypsypops_rubicundus", direction = -1) +
  theme_bw()
```

    ## Warning: Removed 12 rows containing missing values (geom_point).

<img src="README_files/figure-markdown_github/tldr_ggplot-2.png" width="672" />

``` r
# get ecological information from FishBase
data <- rfishbase::ecology(gsub("_"," ", spp), c("SpecCode","FeedingType", "DietTroph")) %>% 
  dplyr::left_join( rfishbase::species(gsub("_"," ", spp)))

# plot boxplots of length across feeding groups using discrete colors provided by Cirrilabrus solorensis
ggplot(data[!is.na(data$FeedingType),]) +
  geom_boxplot(aes(x = FeedingType, y = log(Length), fill = FeedingType )) +
  scale_fill_fish_d(option = "Cirrhilabrus_solorensis", labels = c("invertivore", "herbivore", "carnivore", "planktivore", "omnivore")) +
  theme_bw() +
  theme(axis.text.x= element_blank() )
```

<img src="README_files/figure-markdown_github/tldr_ggplot-3.png" width="672" />

``` r
# examine relationships between size and trophic level with vulnerability as a continuous color scheme provided by Lepomis megalotis
ggplot(data) +
  geom_point(aes(x = Length, y = DietTroph, color = Vulnerability), size = 6, alpha = 0.9) +
  scale_color_fish(option = "Lepomis_megalotis", direction = -1) +
  theme_bw()
```

    ## Warning: Removed 192 rows containing missing values (geom_point).

<img src="README_files/figure-markdown_github/tldr_ggplot-4.png" width="672" />

Colors can also be used with maps. Here are several examples of discrete
and continuous color schemes on a world-map.

``` r
library(ggplot2)

#get dataset of the world's countries
world <- rnaturalearth::ne_countries(scale = "medium", returnclass = "sf")

#plot worldmap with each country's estimated population as a continuous colors scale based on the reverse colors of Whitley's Boxfish Ostracion whitleyi
ggplot(data = world) +
  geom_sf(aes(fill = pop_est)) +
  scale_fill_fish(option = "Ostracion_whitleyi", direction = -1) +
  theme_bw() +
  theme(legend.position = "top")
```

<img src="README_files/figure-markdown_github/ggplot2-1.png" width="672" />

``` r
#plot worldmap with each country's estimated gdp based on the colors of the Sailfin Tang Zebrasoma velifer
ggplot(data = world) +
  geom_sf(aes(fill = gdp_md_est)) +
  scale_fill_fish(option = "Zebrasoma_velifer", trans = "sqrt") +
  theme_bw()+
  theme(legend.position = "top")
```

<img src="README_files/figure-markdown_github/ggplot2-2.png" width="672" />

``` r
#same example as above but starting at a lighter point of the color scale
ggplot(data = world) +
  geom_sf(aes(fill = gdp_md_est)) +
  scale_fill_fish(option = "Zebrasoma_velifer", trans = "sqrt", begin = 0.3, end = 1) +
  theme_bw()+
  theme(legend.position = "top")
```

<img src="README_files/figure-markdown_github/ggplot2-3.png" width="672" />

``` r
#plot worldmap again, this time with countries colored by their respective regional affiliation using the colors of the Clown coris *Coris gaimard* and 'discrete = TRUE'
ggplot(data = world) +
  geom_sf(aes(fill = region_wb)) +
  scale_fill_fish(option = "Coris_gaimard", discrete = TRUE) +
  theme_bw()+
  theme(legend.position = "top")
```

<img src="README_files/figure-markdown_github/ggplot2-4.png" width="672" />

``` r
##same map with colors reversed
ggplot(data = world) +
  geom_sf(aes(fill = region_wb)) +
  scale_fill_fish(option = "Coris_gaimard", discrete = TRUE, direction = -1) +
  theme_bw()+
  theme(legend.position = "top")
```

<img src="README_files/figure-markdown_github/ggplot2-5.png" width="672" />

``` r
#another map with countries colored by economic status using the colors of the Mandarinfish *Synchiropus splendidus*
ggplot(data = world) +
  geom_sf(aes(fill = income_grp)) +
  scale_fill_fish(option = "Synchiropus_splendidus", discrete = T, alpha = 0.8) +
  theme_bw()+
  theme(legend.position = "top")
```

<img src="README_files/figure-markdown_github/ggplot2-6.png" width="672" />

``` r
#same map as above but with a narrower color palette in discrete values
ggplot(data = world) +
  geom_sf(aes(fill = income_grp)) +
  scale_fill_fish(option = "Synchiropus_splendidus", discrete = T, alpha = 0.8, begin = 0.3, end = 1) +
  theme_bw()+
  theme(legend.position = "top")
```

<img src="README_files/figure-markdown_github/ggplot2-7.png" width="672" />

Fish shapes
-----------

fishualize 2.0 enables you to add fish silhouettes to a ggplot object
with the function `add_fishape()`. To get an overview of all available
silhouettes, run `fishapes()`.

``` r
# Default shape
ggplot() + add_fishape() +
  theme_void()
```

<img src="README_files/figure-markdown_github/shapes-1.png" width="672" />

``` r
# Species are specified with family and option
# The color and transparency of the shape is chosen with fill and alpha
# xmin, xmax, ymin, and ymax indicate the position of the fish silhouette
 ggplot(diamonds)+
   geom_bar(aes(cut, fill = cut)) +
   scale_fill_fish_d(option = "Naso_lituratus") +
   add_fishape(family = "Acanthuridae",
               option = "Naso_unicornis",
               xmin = 1, xmax = 3, ymin = 15000, ymax = 20000,
               fill = fish(option = "Naso_lituratus", n = 4)[2],
               alpha = 0.8) +
   theme_bw()
```

<img src="README_files/figure-markdown_github/shapes-2.png" width="672" />

Alternatively, you can specify the coordinates according to a relative
scale (between 0 and 1; with scaled = TRUE). In this case, parameters
xlim and ylim indicate the x and y limits of x and y. This is useful to
keep the fish silhouette positioned identically across graphs with
different scales.

``` r
 ggplot(diamonds)+
   geom_bar(aes(cut, fill = cut)) +
   scale_fill_fish_d(option = "Naso_lituratus") +
   add_fishape(family = "Acanthuridae",
               option = "Naso_unicornis",
               xmin = 0.3, xmax = 0.7, ymin = 0.8, ymax = 1,
               scaled = TRUE,
               xlim = c(0.5, 5.5), ylim = c(0, 21000) ,
               fill = fish(option = "Naso_lituratus", n = 5)[3],
               alpha = 1) +
   theme_bw()
```

<img src="README_files/figure-markdown_github/shapes relative-1.png" width="672" />

Contribute
----------

Love it? Missing your favorite species? Check out how you can contribute
to this package
[here](https://nschiett.github.io/fishualize/articles/contribute.html)

Citation
--------

To cite package `fishualize` in publications use:

Nina Schiettekatte, Simon Brandl and Jordan Casey (2019). fishualize:
Color Palettes Based on Fish Species. R package version 0.1.0.
<https://CRAN.R-project.org/package=fishualize>

A BibTeX entry for LaTeX users is

@Manual{, title = {fishualize: Color Palettes Based on Fish Species},
author = {Nina Schiettekatte and Simon Brandl and Jordan Casey}, year =
{2019}, note = {R package version 0.1.0}, url =
{<https://CRAN.R-project.org/package=fishualize>}, }

Contributions
-------------

JMC, SJB, and NMDS created the color palettes, SJB drew the fish
silhouettes, which are freely available
[**here**](https://github.com/simonjbrandl/fishape/tree/master/shapes).
NMDS wrote the functions. SJB, JMC, and NMDS wrote the documentation. A
big thanks to everyone that submitted additional color palettes via
GitHub: Jindra Lacko, Andrew Steinkruger, Adam Smit, pomboal, anhsmith,
Kent Sorgon, Jean-Pierre Rossi, jdittrich, amedelin, Leonardo Donato
Nunes, tesaucier, Thais Araújo, Áthila Bertoncini Andrade

Credits
-------

The initial structure of this package was inspired by the `harrypotter`
package by Alejandro Jiménez: <https://github.com/aljrico/harrypotter>


NEWS.md
# fishualize 0.2.0

* Added a `NEWS.md` file to track changes to the package.

* Added 17 new fish color palettes. In total there are now 101 palettes.

* Added a function `add_fishape()`

* Added a function `fishapes()` to get an overview of the available fish silhouettes.
There are 101 available fish shapes from 61 families. 

# fishualize 0.2.1

* Fixed CRAN policy issue:
'Packages which use Internet resources should fail gracefully with an informative message if the resource is not available or has changed (and not give a check warning nor error).'
More color palettes.      

# fishualize 0.2.2

* Fixed CRAN policy issue:
'Packages which use Internet resources should fail gracefully with an informative message if the resource is not available or has changed (and not give a check warning nor error).'
--> added an extra statement to test if object shapes is null due to bad internet connection   

# fishualize 0.2.3

* Fixed CRAN policy issue:
'Packages which use Internet resources should fail gracefully with an informative message if the resource is not available or has changed (and not give a check warning nor error).'
--> corrected wrong statement about internet connection


R/fish_pallette.R
#' Original fish color database
#'
#' A dataset containing some colour palettes inspired by fish species
#'
#'
#'@format A data frame containing all the colours used in the palette:
#'\itemize{
#'   \item option: It is intended to be a general option for choosing the specific colour palette.
#'   \item hex: hex color code
#'}
"fishcolors"



#' Available Palettes.
#'
#' This function returns a vector containing the names of all the available palettes in the 'fishualize' package.
#'
#' @return \code{fish_palettes} returns a character vector with the names of the fish palettes available to use.
#'
#'
#'
#' @examples
#' fish_palettes()
#'
#' @rdname fish_palettes
#' @export

fish_palettes <- function(){
  return(sort(unique(fishualize::fishcolors$option)))
}

#'
#'
#'
#'
#' fish Colour Map.
#'
#' This function creates a vector of \code{n} equally spaced colors along the
#' 'fish colour map' of your selection
#'
#' @param n The number of colors (\eqn{\ge 1}) to be in the palette.
#'
#' @param alpha	The alpha transparency, a number in [0,1], see argument alpha in
#' \code{\link[grDevices]{hsv}}.
#'
#' @param begin The (corrected) hue in [0,1] at which the fish colormap begins.
#'
#' @param end The (corrected) hue in [0,1] at which the fish colormap ends.
#'
#' @param direction Sets the order of colors in the scale. If 1, the default, colors
#' are ordered from darkest to lightest. If -1, the order of colors is reversed.
#'
#' @param option A character string indicating the fish species to use.
#'
#'
#' @return \code{fish} returns a character vector, \code{cv}, of color hex
#' codes. This can be used either to create a user-defined color palette for
#' subsequent graphics by \code{palette(cv)}, a \code{col =} specification in
#' graphics functions or in \code{par}.
#'
#'
#'
#' Semi-transparent colors (\eqn{0 < alpha < 1}) are supported only on some
#' devices: see \code{\link[grDevices]{rgb}}.
#'
#' @examples
#' library(ggplot2)
#' library(fishualize)
#'
#' dat <- data.frame(x = rnorm(1e4), y = rnorm(1e4))
#' ggplot(dat, aes(x = x, y = y)) +
#'   stat_density_2d(geom = "raster",
#'   aes(fill = after_stat(density)), contour = FALSE) +
#'   scale_fill_gradientn(colors = fish(128, option = 'Ostracion_cubicus'))
#'
#' pal <- fish(256, option = "Thalassoma_hardwicke", direction = -1)
#' image(volcano, col = pal)
#'
#' @rdname fish
#' @export
#'
fish <- function(n, alpha = 1, begin = 0, end = 1, direction = 1,
                 option = "Centropyge_loricula") {

  if (!option %in% fishualize::fishcolors$option) {
    stop("Unknown, or possibly misspelled fish species.")
  }


  if (begin < 0 | begin > 1 | end < 0 | end > 1) {
    stop("begin and end must be in [0,1]")
  }

  if (abs(direction) != 1) {
    stop("direction must be 1 or -1")
  }

  if (direction == -1) {
    tmp <- begin
    begin <- end
    end <- tmp
  }

  map <- fishualize::fishcolors[fishualize::fishcolors$option == option, ]

  map_cols <- map$hex
  fn_cols <- grDevices::colorRamp(map_cols, space = "Lab",
                                  interpolate = "spline")
  cols <- fn_cols(seq(begin, end, length.out = n)) / 255
  grDevices::rgb(cols[, 1], cols[, 2], cols[, 3], alpha = alpha)
}


#' @rdname fish
#'
#' @export
fish_pal <- function(alpha = 1, begin = 0, end = 1, direction = 1,
                     option = "Centropyge_loricula") {


  function(n) {
    fish(n, alpha, begin, end, direction, option)
  }
}

#' Visualization of fish color palette
#'
#' This function creates an image of the specified fish color palette.
#'
#' @param option A character string indicating the fish species to use.
#'
#' @param n The number of colors (\eqn{\ge 1}) to be in the palette.
#'
#' @param ... Other arguments as can be specified in the function \code{fish}.
#' See ?fishualize::fish for details.
#'
#'
#' @return \code{fishualize} returns a visualisation of the specified color palette.
#'
#' @examples
#'
#' fishualize::fishualize()
#' fishualize::fishualize(option = "Zanclus_cornutus", n = 8)
#'
#' @rdname fishualize
#' @importFrom graphics image
#' @export
#'

fishualize <- function(option = "Centropyge_loricula", n = 5, ...) {
  col <- fishualize::fish(n = n, option = option, ...)
  image(1:n, 1, as.matrix(1:n), col = col,
        main = "", ylab = "", xlab = " ", xaxt = "n", yaxt = "n",  bty = "n")
}


#' @rdname scale_fish
#'
#' @importFrom ggplot2 scale_fill_gradientn scale_color_gradientn discrete_scale
#'
#' @export
scale_color_fish <- function(option = "Centropyge_loricula", ...,
                             alpha = 1, begin = 0, end = 1, direction = 1,
                             discrete = FALSE) {

  if (discrete) {
    discrete_scale("colour", "fish",
                   fish_pal(alpha, begin, end, direction, option), ...)
  } else {
    scale_color_gradientn(colours =
              fish(256, alpha, begin, end, direction, option), ...)
  }
}

#' @rdname scale_fish
#' @aliases scale_color_fish
#' @importFrom ggplot2 discrete_scale
#' @export
scale_colour_fish <- scale_color_fish

#' @rdname scale_fish
#' @aliases scale_color_fish
#' @export
scale_colour_fish_d <- function(option = "Centropyge_loricula", ...,
                                alpha = 1, begin = 0, end = 1,
                                direction = 1) {
  discrete_scale("colour", "fish",
                 fish_pal(alpha, begin, end, direction, option), ...)
}

#' @rdname scale_fish
#' @aliases scale_color_fish
#' @export
scale_color_fish_d <- scale_colour_fish_d


#' @rdname scale_fish
#' @aliases scale_fill_fish
#' @importFrom ggplot2 discrete_scale
#' @export
scale_fill_fish_d <- function(option = "Centropyge_loricula", ...,
                              alpha = 1, begin = 0, end = 1,
                              direction = 1) {
  discrete_scale("fill", "fish",
                 fish_pal(alpha, begin, end, direction, option), ...)
}


#' fish colour scales
#'
#' Uses the fish color scale.
#'
#' For \code{discrete == FALSE} (the default) all other arguments are as to
#' \link[ggplot2]{scale_fill_gradientn} or \link[ggplot2]{scale_color_gradientn}.
#' Otherwise the function will return a \code{discrete_scale} with the plot-computed
#' number of colors.
#'
#'
#' @param ... parameters to \code{discrete_scale} or \code{scale_fill_gradientn}
#'
#' @param alpha pass through parameter to \code{fish}
#'
#' @param begin The (corrected) hue in [0,1] at which the fish colormap begins.
#'
#' @param end The (corrected) hue in [0,1] at which the fish colormap ends.
#'
#' @param direction Sets the order of colors in the scale. If 1, the default, colors
#' are as output by \code{fish_pal}. If -1, the order of colors is reversed.
#'
#' @param discrete generate a discrete palette? (default: \code{FALSE} - generate continuous palette)
#'
#' @param option A character string indicating the fish species to use.
#'
#' @rdname scale_fish
#'
#'
#' @importFrom ggplot2 scale_fill_gradientn scale_color_gradientn discrete_scale
#'
#' @importFrom gridExtra grid.arrange
#'
#' @examples
#' library(ggplot2)
#' library(fishualize)
#'
#'
#' ggplot(diamonds, aes(factor(cut), fill=factor(cut))) +
#' geom_bar() +
#' scale_fill_fish(discrete = TRUE, option = "Centropyge_loricula")
#'
#' ggplot(mtcars, aes(factor(gear), fill=factor(carb))) +
#' geom_bar() +
#' scale_fill_fish(discrete = TRUE, option = "Trimma_lantana")
#'
#' ggplot(mtcars, aes(x = mpg, y = disp, colour = drat)) +
#' geom_point(size = 4) +
#' scale_colour_fish(option = "Ostracion_cubicus", direction = -1)
#'
#'
#'
#'
#' @export
scale_fill_fish <- function(option = "Centropyge_loricula", ...,
                            alpha = 1, begin = 0, end = 1, direction = 1,
                            discrete = FALSE) {

  if (discrete) {
    discrete_scale("fill", "fish",
            fish_pal(alpha, begin, end, direction, option), ...)
  } else {
    scale_fill_gradientn(colours =
            fish(256, alpha, begin, end, direction, option), ...)
  }}


R/fish_shapes.R
try_GET <- function(x, ...) {
  tryCatch(
    httr::GET(url = x, httr::timeout(10), ...),
    error = function(e) conditionMessage(e),
    warning = function(w) conditionMessage(w)
  )
}
is_response <- function(x) {
  class(x) == "response"
}


#' Available fish silhouettes
#'
#' This function returns a dataframe containing the all the available fish
#' silhouettes accessible through the 'fishualize' package.
#'
#' @return \code{fishapes} returns a dataframe containing the all the available fish
#' silhouettes available to use.
#'
#' @importFrom httr GET
#' @importFrom httr stop_for_status
#' @importFrom httr content
#' @importFrom curl has_internet
#' @importFrom httr message_for_status
#' @importFrom magrittr "%>%"
#' @importFrom stringr str_split
#' @importFrom stringr str_subset
#' @importFrom stringr str_sub
#' @importFrom stringr str_replace
#' @importFrom tidyr separate
#' @importFrom dplyr mutate
#' @importFrom rlang .data
#'
#'
#' @examples
#' fishapes()
#'
#' @rdname fishapes
#' @export



fishapes <- function(){

  # First check internet connection
  if (!curl::has_internet()) {
    message("No internet connection.")
    return(invisible(NULL))
  }

  url <- "https://github.com/simonjbrandl/fishape/tree/master/shapes"

  # Then try for timeout problems
  resp <- try_GET(url)
  if (!is_response(resp)) {
    message(resp)
    return(invisible(NULL))
  }
  # Then stop if status > 400
  if (httr::http_error(resp)) {
    httr::message_for_status(resp)
    return(invisible(NULL))
  }
  req <- httr::GET(url)

  httr::stop_for_status(req)
  text <- httr::content(req, "text")

  text_sub <- stringr::str_split(text, "\\s+")[[1]] %>%
    stringr::str_subset(".png")  %>%
    stringr::str_subset("title") %>%
    stringr::str_sub(start = 8, end = -6)

  df <- data.frame(text_sub = text_sub) %>%
    tidyr::separate(text_sub, into = c("family", "option"), sep = "_") %>%
    dplyr::mutate(option = stringr::str_replace(.data$option, "[.]", "_"))

  df
}





#' fish silhouette in ggplot2
#'
#' Adds a fish silhouette to your plot
#'
#' @param family character string indicating the fish family.
#' @param option character string indicating the fish species.
#' If NA, the first available option within a family will be selected
#' @param xmin x location giving minimum horizontal location of silhouette
#' @param xmax x location giving maximum horizontal location of silhouette
#' @param ymin y location giving minimum vertical location of silhouette
#' @param ymax y location giving maximum vertical location of silhouette
#' @param scaled logical parameter. If TRUE, location parameters
#' (xmin, xmax, ymin, ymax) should range between 0 and 1.
#' If FALSE, location parameters should be provided according to the values on
#' the plot axes.
#' @param xlim,ylim vectors of length = 2, contains the data limits
#' and must be provided if scaled is TRUE.
#' @param fill color of fish shape
#' @param alpha transparency of fish shape (value between 0 and 1)
#'
#' @rdname add_fishape
#'
#'
#' @importFrom ggplot2 annotation_custom
#' @importFrom grid rasterGrob
#' @importFrom png readPNG
#' @importFrom scales alpha
#' @importFrom httr GET
#' @importFrom httr message_for_status
#' @importFrom curl has_internet
#'
#'
#' @examples
#' library(ggplot2)
#'
#' ggplot() + add_fishape(fill = fish(n = 5)[4])
#'
#' ggplot(diamonds)+
#'   geom_bar(aes(cut, fill = cut)) +
#'   scale_fill_fish_d(option = "Naso_lituratus") +
#'   add_fishape(family = "Acanthuridae",
#'               option = "Naso_unicornis",
#'               xmin = 1, xmax = 3, ymin = 15000, ymax = 20000,
#'               fill = fish(option = "Naso_lituratus", n = 5)[3],
#'               alpha = 0.8) +
#'   theme_bw()
#'
#' ## example with relative coordinates
#' ggplot(diamonds)+
#'   geom_bar(aes(cut, fill = cut)) +
#'   scale_fill_fish_d(option = "Naso_lituratus") +
#'   add_fishape(family = "Acanthuridae",
#'               option = "Naso_unicornis",
#'               xmin = 0, xmax = 0.3, ymin = 0.8, ymax = 1,
#'               scaled = TRUE,
#'               xlim = c(0.5, 5.5), ylim = c(0, 21000) ,
#'               fill = fish(option = "Naso_lituratus", n = 5)[3],
#'               alpha = 1) +
#'   theme_bw()
#'
#' @export
#'
add_fishape <- function(family = "Pomacanthidae",
                        option = "Centropyge_loricula",
                        xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = Inf,
                        scaled = FALSE,
                        xlim = NULL, ylim = NULL,
                        fill = "#000000",
                        alpha = 1){

  # check if shape is available

  shapes <- fishapes()

  if (is.na(option)){
    option <- shapes[shapes$family == family, "option"][1]
  }

  if (nrow(shapes[shapes$family == family, ]) == 0){
    stop("This family is not available or misspelled")
  }

  if (nrow(shapes[shapes$option == option, ]) == 0){
    stop("This species option is not available or misspelled")
  }

  # get shape

  url <- paste0(
    "https://raw.githubusercontent.com/simonjbrandl/fishape/master/shapes/",
    family, "_", gsub("_", ".", option), ".png")

  try_GET <- function(x, ...) {
    tryCatch(
      httr::GET(url = x, httr::timeout(10), ...),
      error = function(e) conditionMessage(e),
      warning = function(w) conditionMessage(w)
    )
  }

  is_response <- function(x) {
    class(x) == "response"
  }

  # First check internet connection
  if (!curl::has_internet()) {
    message("No internet connection.")
    return(invisible(NULL))
  }
  # Then try for timeout problems
  resp <- try_GET(url)
  if (!is_response(resp)) {
    message(resp)
    return(invisible(NULL))
  }
  # Then stop if status > 400
  if (httr::http_error(resp)) {
    httr::message_for_status(resp)
    return(invisible(NULL))
  }


  img <- wrap.url(url, png::readPNG)
  g <- grid::rasterGrob(img, interpolate=TRUE)

  # reset color and alpha
  oldcol <- names(sort(table(g$raster), decreasing=TRUE))
  oldcol <- oldcol[!oldcol == "#FFFFFF00"]
  oldcol <- oldcol[1]

  newcol <- scales::alpha(fill, alpha)
  g$raster[g$raster != oldcol] <- NA
  g$raster[g$raster == oldcol] <- newcol

  # possibility to rescale
  if (scaled == TRUE){
    ## first check if all info is there
    if(xmin < 0 | xmax > 1 | ymin < 0 | ymax > 1 |
       is.null(xlim) | is.null(ylim)){
      stop("If scaled = TRUE,
           min and max of x and y have to lie between 0 and 1 and
           xlim and ylim have to be defined")
    }
    ## rescale to absolute
    xmin2 <- ((xlim[2] - xlim[1]) * xmin) + xlim[1]
    xmax2 <- ((xlim[2] - xlim[1]) * xmax) + xlim[1]
    ymin2 <- ((ylim[2] - ylim[1]) * ymin) + ylim[1]
    ymax2 <- ((ylim[2] - ylim[1]) * ymax) + ylim[1]

    ggplot2::annotation_custom(g, xmin = xmin2, xmax = xmax2,
                               ymin = ymin2, ymax = ymax2)

  } else{
    ggplot2::annotation_custom(g, xmin = xmin, xmax = xmax,
                               ymin = ymin, ymax = ymax)
  }

}

###### function borrowed from r package imager #####
wrap.url <- function(file,fun)
{
  is.url <- grepl("^(http|ftp)s?://", file)
  if (is.url)
  {
    url <- file
    ext <- stringr::str_extract_all(url,"\\.([A-Za-z0-9]+$)")[[1]]
    if (length(ext) > 0) file <- tempfile(fileext=ext) else file <- tempfile()
    downloader::download(url,file,mode="wb")
    out <- fun(file)
    unlink(file)
    out
  }
  else
  {
    if (!utils::file_test("-f",file))
    {
      stop("File not found")
    }
    else
    {
      fun(file)
    }
  }

}




